### PR TITLE
Abstract `remappings.txt` generation into `sol_remappings` rule.

### PR DESCRIPTION
**Pull request by aschlosberg**
In using the previously introduced `remappings` attribute of `sol_binary` we realised that its name was confusing (suggesting input, not output) and that it was too limiting in its functionality. The new `SolRemappingsInfo` provider acts as a "combiner" of `remappings` attributes and the `sol_remappings` rule replaces the file generation previously performed by `sol_binary`. See `e2e/workspace/BUILD` for example usage.

### CODE REVIEW

File: sol/private/common.bzl [hunk 1]
```diff
@@ -0,0 +1,28 @@
+"""Shared functionality used by one or more rules."""
+
+load("//sol:providers.bzl", "SolRemappingsInfo")
+
+def transitive_remappings(ctx, extra_remappings = {}):
```
Comment (alexeagle) [inline sol/private/common.bzl:5]: opt: everywhere this is used, it's wrapped in a SolRemappingsInfo. Maybe it wants to be a factory function that creates SolRemappingsInfo, taking deps into account, similar to the factory function for js_info: https://github.com/aspect-build/rules_js/blob/main/js/private/js_info.bzl#L18
Then at the use sites, you'd construct a SolRemappingsInfo, and you can still iterate over the dict to format the entries as command-line args or a remappings.txt file

File: sol/providers.bzl [hunk 2]
```diff
@@ -5,6 +5,17 @@ SolSourcesInfo = provider(
...
         "transitive_sources": "depset of transitive dependency sources",
+    },
+)
+
+SolRemappingsInfo = provider(
+    doc = """Stores a dictionary of solc remappings.
+    
+    Allows for piping of remappings through a dependency tree of targets. Rules
+    that accept a "remappings" attribute and/or dependencies that provide
+    SolRemappingsInfo SHOULD propagate their union via a SolRemappingsInfo.
```
Comment (alexeagle) [inline sol/providers.bzl:16]: if you use the factory function, then rules should never construct a `SolRemappingsInfo` themselves, which makes it harder for them to get the propagation wrong. You can make the SolRemappingsInfo symbol only available in the private/ folder, then make the factory function the public API instead

**Review by alexeagle — APPROVED** [review]
cool

File: sol/private/common.bzl (hunk 1 above)
Comment (aschlosberg) [reply sol/private/common.bzl]: Good catch, thank you. Done.

File: sol/providers.bzl (hunk 2 above)
Comment (aschlosberg) [reply sol/providers.bzl]: Done.

**Review by alexeagle — APPROVED** [review]

**Pull request merged**